### PR TITLE
Enable threaded calculations with thermochimica

### LINKED ISSUE #25969: Enable threaded calculations with thermochimica
## Reason
Thermodynamic solves with Thermochimica make up a large fraction of coupled calculations. We'd like to speed up the overall simulations by taking advantage of threaded calculations (where for example we are using a direct solver for the non-linear system, which does not scale as well). However Thermiochimica uses tons of global variables and is thus not thread-safe.

## Design
We could `fork()` a child process for every thread and exchange data using shared memory (`mmap` with `MAP_SHARED`). Then we use bidirectional sockets to send notification messages to syncronize the processes.

## Impact
Speed up

### PR DESCRIPTION
**Pull request by dschwen**
Depends on #25880

Closes #25969

### CODE REVIEW

**Comment by dschwen** [comment]
Ok, this is curious, the runs that failed timed out, but they all finished. I'm wondering if civet waits for the child processes. On Mac they seem to get killed with the parent process exiting.

**Comment by loganharbour** [comment]
> Ok, this is curious, the runs that failed timed out, but they all finished. I'm wondering if civet waits for the child processes. On Mac they seem to get killed with the parent process exiting.

It's not in civet, it's in the test harness. the test harness reports the full running time that civet also reports.

**Comment by dschwen** [comment]
I think I need to explicitly kill the child process in the destructor. Otherwise it will hang around as a zombie (child of init).

**Comment by dschwen** [comment]
This is hilarious. Since the thermochimica server is running in a child process the code coverage doesn't see it _at all_. @loganharbour

**Comment by dschwen** [comment]
Ok, I think I know what's going on. I used `SIGKILL` to terminate the child process. That bypasses any _at exit_ cleanup , which includes a `_gcov_flush()`. I changed this to a `SIGTERM`, which should allow the child process to write the gcov data before terminating.

**Comment by dschwen** [comment]
Hm, the coverage went up big-time, but the diff was not updated above.

**Comment by dschwen** [comment]
@loganharbour , why does the coverage diff above not get updated?

File: modules/chemical_reactions/src/userobjects/ThermochimicaData.C [hunk 1]
```diff
@@ -146,6 +154,58 @@ ThermochimicaDataBase<is_nodal>::ThermochimicaDataBase(const InputParameters & p
...
+  // set up shared memory for communication with child process
+  auto shared_mem =
+      static_cast<std::byte *>(mmap(nullptr,
+                                    dofid_size * sizeof(dof_id_type) + real_size * sizeof(Real),
+                                    PROT_READ | PROT_WRITE,
+                                    MAP_ANONYMOUS | MAP_SHARED,
+                                    -1 /* fd */,
+                                    0 /* offset */));
+  if (shared_mem == MAP_FAILED)
+    mooseError(
```
Comment (loganharbour) [inline modules/chemical_reactions/src/userobjects/ThermochimicaData.C:173]: All of these errors will be prefixed by the object type and name, so your context is extraneous

**Review by loganharbour — APPROVED** [review]
Looks good. Not a bad idea to have @parikshitbajpai give a final glance.

**Review by parikshitbajpai — APPROVED** [review]
This looks good to me

File: modules/chemical_reactions/src/userobjects/ThermochimicaData.C (hunk 1 above)
Comment (dschwen) [reply modules/chemical_reactions/src/userobjects/ThermochimicaData.C:173]: ah right, we changed that for the MooseServer, right?

**Pull request merged**